server/cluster/unsafe_recovery_controller.go

```diff
@@ -128,9 +130,111 @@ func (u *unsafeRecoveryController) isPlanExecuted(report *pdpb.StoreReport) bool
 	return true
 }
 
+type regionItem struct {
+	region *metapb.Region
+}
+
+func (r regionItem) Less(other btree.Item) bool {
+	return bytes.Compare(r.region.StartKey, other.(regionItem).region.StartKey) < 0
+}
+
+func (u *unsafeRecoveryController) canElectLeader(region *metapb.Region) {
+	numFailedPeers := 0
+	for _, peer := range reigon.Peers {
+		_, ok := u.failedStores[peer.StoreId]
+		if ok {
+			numFailedPeers += 1
+		}
+	}
+	return numFailedPeers*2 < len(region.Peers)
+}
+
 func (u *unsafeRecoveryController) generateRecoveryPlan() {
 	u.Lock()
 	defer u.Unlock()
+	// Peers that cannot elect a leader due to store failures.
+	leaderlessPeers := make([]*pdpb.PeerReport)
+	availableRegionReports := make(map[uint64]*metapb.Region)
+	for _, storeReport := range u.storeReports {
+		for _, peerReport := range storeReports.Reports {
+			region := peerReport.RegionState.Region
+			if u.canElectLeader(region) {
+				existing, ok := availableRegionReports[region.Id]
+				if ok && existing.RegionEpoch.ConfVer > region.RegionEpoch.ConfVer && existing.RegionEpoch.Version > region.RegionEpoch.Version {
+					// Only keeps the newest region report.
+					continue
+				}
+				availableRegionReports[region.Id] = region
+			} else {
+				leaderlessPeers = append(leaderlessPeers, peerReport)
+			}
+		}
+	}
+	// Uses 2 ordered map to implement the interval map.
+	validRegions := btree.New(2)
+	for _, region := range regionReports {
+		validRegions.ReplaceOrInsert(regionItem{region})
+	}
+	sort.SliceTable(leaderlessPeers, func(i, j int) bool {
+		return leaderpeerlessPeers[i].RaftState.LastIndex > leaderlessPeers[j].RaftState.LastIndex
+	})
+	for _, peer := range leaderlessPeers {
+		// Iterates all leaderless peers, if a peer has been fully
+		// covered by a valid range (from another region or the same region but another peer), do nothing, while, if a peer can fill some
+		// of the gaps between available ranges so far, find all the gaps and ask the peer
+		// to create new regions for each of the gap, besides, merge the all the scattered
+		// ranges that are connected by filling up these gaps.
+		startKey := peer.RegionState.Region.StartKey
+		endKey := peer.RegionState.Region.EndKey
+		overlapRegions := make([]*metapb.Region, 0)
+		validRegions.DescendLessOrEqual(regionItem{&metapb.Region{StartKey: startKey}}, func(item btree.Item) bool {
+			if item.(regionItem).region.StartKey != startKey {
+				// If the start keys are the same, the region will still be added while we
+				// traverse the btree in ascending order below.
+				overlapRegions = append(overlapRegions, item.(regionItem).region)
```

Comment:
Since the btree is order by the start key, it is possible that the region found by this "search" overlaps with the leaderless region, and, as you said, it is also possible that they do not overlap with each other. We do not know this until we compare the end keys which I did several lines below. It is just a implementation choice. Does not affect the result I think